File: Languages/Typescript/Algorithms/binarySearch.ts
console.log("my binary search first try");

let testArr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
let testKeys = Object.keys(testArr);

type InputType = {
  values: number[];
  keys: string[];
};

function conquer(left: InputType, right: InputType, target: number): InputType {
  if (right.values[0] <= target) {
    if (right.values[0] === target) {
      return { values: [right.values[0]], keys: [right.keys[0]] };
    }
    return { values: right.values, keys: right.keys };
  } else {
    if (left.values[0] === target) {
      return { values: [left.values[0]], keys: [left.keys[0]] };
    }

    return { values: left.values, keys: left.keys };
  }
}

function binarySearch(arr: InputType, target: number) {
  if (arr.values.length <= 1) {
    console.log(arr);
    return [arr];
  }

  const mid = arr.values.length / 2;

  const left = arr.values.splice(0, mid);
  const leftKeys = arr.keys.splice(0, mid);

  binarySearch(conquer({ values: left, keys: leftKeys }, arr, target), target);
}

console.time("start");
binarySearch({ values: testArr, keys: testKeys }, Number(7));
console.timeEnd("start");

// { values: [ 7 ], keys: [ '6' ] }

File: Languages/Java/Algorithms/BinarySearch.java
class Algorithms {
  
    private static int binarySearch(int left, int right, List<Integer> arr, int target){
        int mid = left + (right-left) /2;
        if(arr.get(mid) == target) return mid;
        if(arr.get(mid) >target) return binarySearch(left, mid-1, arr, target); 
        if(arr.get(mid)<target) return binarySearch(mid+1, right, arr, target);
        return -1; // not found
        
    }

    public static int getIndexWithBinarySearch(int target, List<Integer> arr) {
        int n = arr.size();
        int left = 0;
        int right = n-1;
        
        return binarySearch(left, right, arr, target);
    }

}
